ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -216,6 +255,102 @@ KEXT_STATIC void UseMainForkIfNamedStream(
     }
 }
 
+static void ResizePendingRenames(uint32_t newMaxPendingRenames)
+{
+    PendingRenameOperation* newArray = Memory_AllocArray<PendingRenameOperation>(newMaxPendingRenames);
+    PendingRenameOperation* arrayToFree = nullptr;
+    uint32_t arrayToFreeLength = 0;
+    
+    SpinLock_Acquire(s_renameLock);
+    {
+        if (newMaxPendingRenames >= s_maxPendingRenames)
+        {
+            Array_CopyElements(newArray, s_pendingRenames, s_maxPendingRenames);
+            Array_DefaultInit(newArray + s_maxPendingRenames, newMaxPendingRenames - s_maxPendingRenames);
+            
+            arrayToFree = s_pendingRenames;
+            arrayToFreeLength = s_maxPendingRenames;
+            
+            s_pendingRenames = newArray;
+            s_maxPendingRenames = newMaxPendingRenames;
+        }
+    }
+    SpinLock_Release(s_renameLock);
+    
+    if (arrayToFree != nullptr)
+    {
+        assert(arrayToFreeLength > 0);
+        Memory_FreeArray(arrayToFree, arrayToFreeLength);
+    }
+}
+
+static void RecordPendingRenameOperation(vnode_t vnode)
+{
+    thread_t myThread = current_thread();
+    
+    bool resizeTable;
+    do
+    {
+        resizeTable = false;
+        uint32_t resizeTableLength = 0;
+        
+        SpinLock_Acquire(s_renameLock);
+        {
+            bool inserted = false;
+            for (uint32_t i = 0; i < s_maxPendingRenames; ++i)
+            {
+                if (s_pendingRenames[i].thread == nullptr)
+                {
+                    s_pendingRenames[i].thread = myThread;
+                    s_pendingRenames[i].vnode = vnode;
+                    inserted = true;
+                    break;
+                }
+                else
+                {
+                    assert(s_pendingRenames[i].thread != myThread);
+                }
+            }
+            
+            if (!inserted)
+            {
+                resizeTable = true;
+                resizeTableLength = MAX(s_maxPendingRenames * 2u, 1u);
+            }
+        }
+        SpinLock_Release(s_renameLock);
+        
+        if (resizeTable)
+        {
+            ResizePendingRenames(resizeTableLength);
```

Comment:
Since we are not under lock here, I think 2 threads could have gone thru the previous loop.  Then they would both try to do the resize.  Does that sound correct?

Reply:
Yes, that's something we just have to deal with.

Reply:
Is the issue that we can't resize under lock?

Reply:
Yes, blocking memory allocations are iffy while holding a lock as you risk deadlocking. (There's an in-depth explanation somewhere in #500.) During a spinlock they're a complete no-go.